### ISSUE TITLE
Persistent/Stored XSS in 3.0.0

### ISSUE THREAD

**Issue by x-42** [dradis/dradisframework#116]
There exists what I consider to be a low-risk Stored XSS vulnerability in the web interface. Low-risk mostly because it requires a valid user account, and as far as I know, you can only have one user account. So, prerequisites for successful exploitation would be to own the Dradis user account…

Onto some details…

Dradis allows for the creation of “Nodes”. These Nodes store information regarding any projects you’re working on. These Nodes, additionally allow for “Subnodes”.

There are two different, but related issues I’ve found.

Issue 1, is that javascript is rendered ONLY upon the creation, and deletion of a single top-level node. Not a “stored XSS” vuln. So, adding a top-level node with a label of 
![script_img](https://cloud.githubusercontent.com/assets/10157111/8764374/7d0ff080-2da2-11e5-89d7-0749bb499ea8.png) will execute that script ONLY upon creation and deletion of the node. It will not be rendered by clicking on the node once it’s been added.

Creating the “stored” XSS…

Issue 2, The stored XSS occurs when you add a “subnode” to a previously “XSS-enabled” top-level node you’ve already added.. (any Label title will do). The script that will execute will always be the script that was injected into the top-level node Label.

Steps to reproduce:

1. Add a Node with label: 
![script_img](https://cloud.githubusercontent.com/assets/10157111/8764377/b0247018-2da2-11e5-81fa-eae16f495337.png)
(this can be considered the payload) 

![tln](https://cloud.githubusercontent.com/assets/10157111/8764378/d1524dfa-2da2-11e5-9e5a-70be1e9cb27d.png)

You should see a pop-up with “XSS1″ execute after this step. However, we haven’t “stored” the script yet…

2. Add a “subnode” within that node called “test”. (without the quotes)

![subnode](https://cloud.githubusercontent.com/assets/10157111/8764379/ee3db7f6-2da2-11e5-813e-7c18bb58aaba.png)

(You should also see a pop-up with “XSS1″ execute after this step) This confirms that that our “payload” in step 1 is executed after the addition of this subnode.

3. Stored XSS: Click on the “test” subnode in the left-hand “Nodes” navigation pane to execute the Top-level node javascript. (This is the stored XSS of the top-level node.)

![test](https://cloud.githubusercontent.com/assets/10157111/8764380/0198f252-2da3-11e5-8a56-da513992b30e.png)

![xss](https://cloud.githubusercontent.com/assets/10157111/8764382/0fb5fa1a-2da3-11e5-94b6-902e8664b359.png)

**Comment by etdsoft** [comment]
Hi @x-42,

Thanks for the report. Nothing like a bit of XSS investigation for a slow Sunday.

I'll look into it and report back. If appropriate I'll add some information in our **Security Reports** page at:

http://dradisframework.org/security_reports.html

Also, maybe next time you'd like to give us a heads up privately via **security[ {at} ]dradisframework{ [dot] }org** instead of posting in the public tracker directly.

Any any case, thanks for the report, and watch this space for an update.

**Comment by etdsoft** [comment]
@x-42 are you testing this on the 3.0.0rc2 packages or the git versions?

**Comment by x-42** [comment]
@etdsoft hi. this was tested with dradis-3.0.0.rc2 package.

**Comment by etdsoft** [comment]
Ok, I've got rc3 ready, but nowhere to upload them, what architecture you need? (macos, linux:86 or linux:86_64)?

**Comment by x-42** [comment]
actually, i was using the 32 bit version in a 64 bit linux kali install now that i think about it. i can test the rc2 linux 64 bit version in a bit if you need me to.

**Comment by etdsoft** [comment]
No, that's fine, `dradis-3.0.0.rc3-linux-x86.tar.gz` 93M, not sure how to pass that to you for testing

**Comment by x-42** [comment]
however works for. i may be able to set you up with a temporary sftp acct. in a bit. let me know. or dropbox or google drive would work for me. whichever works for you.

**Comment by etdsoft** [comment]
Please get it from http://tbf.me/a/BKiUcp

```
$ shasum dradis-3.0.0.rc3-linux-x86.tar.gz
cc04a88baa34dcba97a4f72026266fa015f03ebc  dradis-3.0.0.rc3-linux-x86.tar.gz
```

**Comment by x-42** [comment]
@etdsoft getting an error extracting:

dradis-3.0.0.rc3-linux-x86/lib/vendor/ruby/2.1.0/gems/sinatra-1.4.6/CHANGES
tar: Unexpected EOF in archive
tar: Unexpected EOF in archive
tar: Error is not recoverable: exiting now

i'll try a re-download see if that helps.

**Comment by x-42** [comment]
re-download worked. testing now.

**Comment by x-42** [comment]
issue appears to be fixed. thanks for the quick response on this @etdsoft

**Comment by etdsoft** [comment]
Cool, I'm waiting for CVE numbers, and SF to be up again, then we'll publish the release and  a new note to the Security Reports page.

**Comment by x-42** [comment]
sounds good.. thanks again!
